From 9aa64dfb2dd07a506024218f9d2c533739ba2ac8 Mon Sep 17 00:00:00 2001
From: Muhammad Usama Anjum <usama.anjum@collabora.com>
Date: Sat, 6 Dec 2025 10:31:49 +0500
Subject: [PATCH 038/188] Input: Ignore the KEY_POWER events if hibernation is
 in progress

Input (Serio) drivers call input_handle_event(). Although the serio
drivers have duplicate events, they have separate code path and call
input_handle_event(). Ignore the KEY_POWER such that this event isn't
sent to the userspace if hibernation is in progress.

Abort the hibernation by calling pm_wakeup_dev_event(). In case of
serio,
doesn't have wakeup source registered, this call doesn't do anything.
But there may be other input drivers which will require this.

Without this, the event is sent to the userspace and it suspends the
device after hibernation cancellation.

Signed-off-by: Muhammad Usama Anjum <usama.anjum@collabora.com>
---
 drivers/input/input.c | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/drivers/input/input.c b/drivers/input/input.c
index cf6fecea7..f4fe3d93e 100644
--- a/drivers/input/input.c
+++ b/drivers/input/input.c
@@ -26,6 +26,7 @@
 #include <linux/kstrtox.h>
 #include <linux/mutex.h>
 #include <linux/rcupdate.h>
+#include <linux/suspend.h>
 #include "input-compat.h"
 #include "input-core-private.h"
 #include "input-poller.h"
@@ -362,6 +363,17 @@ void input_handle_event(struct input_dev *dev,
 
 	lockdep_assert_held(&dev->event_lock);
 
+	/*
+	 * During hibernation/suspend, consume KEY_POWER events in the kernel
+	 * to cancel the sleep transition. Don't forward to userspace to prevent
+	 * userspace (e.g., systemd-logind) from triggering another suspend or
+	 * poweroff action after the system wakes up.
+	 */
+	 if (code == KEY_POWER && pm_sleep_transition_in_progress()) {
+		pm_wakeup_dev_event(&dev->dev, 0, true);
+		return;
+	}
+
 	disposition = input_get_disposition(dev, type, code, &value);
 	if (disposition != INPUT_IGNORE_EVENT) {
 		if (type != EV_SYN)
-- 
2.55.0

